tools/testing/e2e/iree-e2e-conv2d-test.cc

```diff
@@ -0,0 +1,557 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <assert.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "iree/base/api.h"
+#include "iree/base/internal/flags.h"
+#include "iree/base/internal/math.h"
+#include "iree/hal/api.h"
+#include "iree/modules/hal/module.h"
+#include "iree/tooling/context_util.h"
+#include "iree/tooling/device_util.h"
+#include "iree/vm/api.h"
+#include "iree/vm/native_module_cc.h"
+#include "tools/testing/e2e/test_utils.h"
+
+//===----------------------------------------------------------------------===//
+// Reference conv2d (NCHW-FCHW)
+//===----------------------------------------------------------------------===//
+
+// Conversion from 4D indices in row major order to 1D index.
+int convert_to_1d_index(iree_hal_dim_t channels, iree_hal_dim_t height,
```

Comment:
Could put these functions in an anonymous namespace too (rather than at the top level)

Reply:
done.